File: test backend/backend/app/routers/ram_products.py
from fastapi import APIRouter, HTTPException
from app.services.ram_products_service import (
    list_rams,
    get_ram_details,
)

router = APIRouter(prefix="/api/products/ram", tags=["RAM"])

@router.get("")
def get_rams(
    memory_type: str | None = None,
    min_capacity: int | None = None,
    min_price: int | None = None,
    max_price: int | None = None,
    limit: int = 24,
    offset: int = 0,
):
    filters = {
        "memory_type": memory_type,
        "min_capacity": min_capacity,
        "min_price": min_price,
        "max_price": max_price,
    }
    return list_rams(filters, limit, offset)


@router.get("/{product_id}")
def get_ram(product_id: int):
    data = get_ram_details(product_id)
    if not data:
        raise HTTPException(404, "RAM not found")
    return data

File: frontend/src/api/ramProducts.js
import { apiFetch } from "./api";

export function fetchRams({
  search,
  memory_type,
  capacity_gb,
  min_price,
  max_price,
  limit = 24,
  offset = 0,
} = {}) {
  const params = new URLSearchParams();

  if (search) params.append("search", search);
  if (memory_type) params.append("memory_type", memory_type);
  if (capacity_gb) params.append("capacity_gb", capacity_gb);
  if (min_price) params.append("min_price", min_price);
  if (max_price) params.append("max_price", max_price);
  params.append("limit", limit);
  params.append("offset", offset);

  return apiFetch(`/api/products/ram?${params.toString()}`);
}

export function fetchRamDetails(productId) {
  return apiFetch(`/api/products/ram/${productId}`);
}
